### PR TITLE
[breaking]Consume ops xref api

### PR DESCRIPTION
**Pull request by live1206**
Pair PR: https://ceapex.visualstudio.com/Engineering/_git/Docs.Build/pullrequest/19194?_a=overview

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/5368)

### CODE REVIEW

File: src/docfx/config/ops/OpsConfigAdapter.cs [hunk 1]
```diff
@@ -114,6 +128,21 @@ public void Dispose()
             _http.Dispose();
         }
 
+        private string GetXrefMapApiEndpoint(string xrefEndpoint)
+        {
+            var isProduction = string.Equals(xrefEndpoint?.TrimEnd('/'), "https://xref.docs.microsoft.com", StringComparison.OrdinalIgnoreCase);
```
Comment (yufeih) [inline src/docfx/config/ops/OpsConfigAdapter.cs]: If `xrefEndPoint` is not set, should it respect the current `DOCS_ENVIRONMENT` setting?

File: src/docfx/config/ops/OpsConfigAdapter.cs [hunk 2]
```diff
@@ -114,6 +128,21 @@ public void Dispose()
...
+            return isProduction
+                    ? "https://op-build-prod.azurewebsites.net"
+                    : "https://op-build-sandbox2.azurewebsites.net";
+        }
+
+        private async Task<string[]> GetXrefMaps(string xrefMapApiEndpoint, string tag, string xrefMapQueryParams)
+        {
+            var url = $"{xrefMapApiEndpoint}/v1/xrefmap{tag}{xrefMapQueryParams}";
+            var response = await Fetch(url, s_opsHeaders);
+            return JsonUtility.Deserialize<XrefMapApiResponse>(response, null).Links;
```
Comment (yufeih) [inline src/docfx/config/ops/OpsConfigAdapter.cs]: nit: to save a class definition:

```suggestion
            return JsonConvert.DeserializeAnonymousType(
```

File: src/docfx/config/ops/OpsConfigAdapter.cs (hunk 1 above)
Comment (live1206) [reply src/docfx/config/ops/OpsConfigAdapter.cs]: yes

File: src/docfx/config/ops/OpsConfigAdapter.cs (hunk 2 above)
Comment (live1206) [reply src/docfx/config/ops/OpsConfigAdapter.cs]: how to describe the class with anonymous type, like `new { new[] { links = [] } }`, which is invalid

Comment (yufeih) [reply src/docfx/config/ops/OpsConfigAdapter.cs]:
```suggestion
            return JsonUtility.DeserializeAnonymousType(response, new { links = new [] { "" } });
```

File: src/docfx/config/ConfigLoader.cs [hunk 3]
```diff
@@ -46,7 +46,7 @@ where glob(configPath)
         /// <summary>
         /// Load the config under <paramref name="docsetPath"/>
         /// </summary>
-        public (List<Error> errors, Config config) Load(string docsetPath, string locale, CommandLineOptions options, bool noFetch = false)
```
Comment (yufeih) [inline src/docfx/config/ConfigLoader.cs]: keep the `locale` here, it is very likely useful in the future

File: src/docfx/config/ConfigLoader.cs [hunk 4]
```diff
@@ -111,14 +111,15 @@ private static JObject LoadConfig(List<Error> errorBuilder, string fileName, str
         }
 
         private JObject DownloadExtendConfig(
-            List<Error> errors, string locale, PreloadConfig config, Repository repository, FileResolver fileResolver)
+            List<Error> errors, PreloadConfig config, JObject opsConfig, Repository repository, FileResolver fileResolver)
```
Comment (yufeih) [inline src/docfx/config/ConfigLoader.cs]: Prefer strong class here. Can make `LoadDocfxConfig` return `(docfxConfig, opsConfig)`

```suggestion
            List<Error> errors, PreloadConfig config, OpsConfig opsConfig, Repository repository, FileResolver fileResolver)
```

File: src/docfx/config/ConfigLoader.cs [hunk 5]
```diff
@@ -111,14 +111,15 @@ private static JObject LoadConfig(List<Error> errorBuilder, string fileName, str
...
             var extendQuery =
-                $"name=" + WebUtility.UrlEncode(config.Name) +
-                $"&locale=" + WebUtility.UrlEncode(locale) +
-                $"&repository_url=" + WebUtility.UrlEncode(repository?.Remote) +
-                $"&branch=" + WebUtility.UrlEncode(repository?.Branch);
+                $"name={WebUtility.UrlEncode(config.Name)}" +
+                $"&repository_url={WebUtility.UrlEncode(repository?.Remote)}" +
+                $"&branch={WebUtility.UrlEncode(repository?.Branch)}" +
+                $"&xref_endpoint={WebUtility.UrlEncode(opsConfig?["xref_endpoint"]?.ToString())}" +
+                $"&xref_query_tags={WebUtility.UrlEncode(string.Join('|', opsConfig?["xref_query_tags"]?.ToString()))}";
```
Comment (yufeih) [inline src/docfx/config/ConfigLoader.cs]: Typically multi items in url are seperated by `,`

File: src/docfx/config/ops/OpsConfigAdapter.cs [hunk 6]
```diff
@@ -106,9 +121,33 @@ private async Task<string> GetBuildConfig(Uri url)
                     Path.Combine(AppContext.BaseDirectory, "data/schemas/OpsMetadata.json"),
                     $"{MetadataSchemaApi}{metadataServiceQueryParams}",
                 },
+                xref = new JArray(xrefMaps),
```
Comment (yufeih) [inline src/docfx/config/ops/OpsConfigAdapter.cs]:
```suggestion
                xref = xrefMaps,
```

File: src/docfx/config/ops/OpsConfigLoader.cs [hunk 7]
```diff
@@ -74,6 +74,8 @@ private static JObject ToDocfxConfig(string branch, OpsConfig opsConfig, PathStr
             {
                 if (!string.IsNullOrEmpty(docsetConfig.DocsetName))
                 {
+                    result["xref_endpoint"] = opsConfig.XrefEndpoint;
+                    result["xref_query_tags"] = string.Join('|', docsetConfig.XrefQueryTags);
```
Comment (yufeih) [inline src/docfx/config/ops/OpsConfigLoader.cs]: are these 2 lines needed? they don't seem to belong to docfx config

File: src/docfx/config/ops/OpsConfigAdapter.cs [hunk 8]
```diff
@@ -106,9 +121,33 @@ private async Task<string> GetBuildConfig(Uri url)
...
             });
         }
 
+        private string GetXrefMapApiEndpoint(string xrefEndpoint)
+        {
+            bool isProduction;
```
Comment (yufeih) [inline src/docfx/config/ops/OpsConfigAdapter.cs]: nit:
```suggestion
            var isProduction = s_isProduction;
            if (!string.IsNullOrEmpty(xrefEndpoint))
                xxx
```

**Pull request merged**